---
title: "Shaded vs exposed ITS2 bioinformatics and analyses"
author: "Robert Lewis"
date: "26/04/2021"
---
Following the dada2 workflow here https://benjjneb.github.io/dada2/tutorial.html, and here https://benjjneb.github.io/dada2/ITS_workflow.html

# Libraries

```{r}
#some packages may require BiocManager for installation 

if (!requireNamespace("BiocManager", quietly = TRUE))
install.packages("BiocManager")
BiocManager::install("dada2")
BiocManager::install("ShortRead")
BiocManager::install("Biostrings")
BiocManager::install("phyloseq")
BiocManager::install("DECIPHER")
BiocManager::install("msa")
BiocManager::install("microbiome")
BiocManager::install("DESeq2")
BiocManager::install("nlstools")


```


```{r}

library(tidyverse)
library(dada2)
library(ShortRead)
library(Biostrings)
library(phyloseq)
library(DECIPHER)
library(msa)
library(phangorn)
library(seqinr)
library(ape) 
library(treemapify)
library(kmer)
library(microbiome)
library(ggpubr)
library(vegan)
library(ggraph)
library(igraph)
library(tidygraph)
library(DESeq2)
library(tidytext)
library(reshape2)
library(bestNormalize)
library(ggrepel)
library(patchwork)
library(car)
library(ggalt)
library(ggVennDiagram)

source("help.R")

```

##load objects
```{r}
ps <- readRDS("ps.RDS")

melt <- readRDS("melt.RDS")

melt_4_deseq2 <- readRDS("melt_4_deseq2.RDS")

ps_sub <- readRDS("ps_sub_w_kmer_tree.rds")

dynamQ_max_EnormEk <- readRDS("dynamQ_max_EnormEk.rds")

FqFmparams <- readRDS("FqFmparams_no_PAR_adjust.rds")

total_brightness <- readRDS("total_brightness.rds")

raro_metadata <- read.csv("metadata_complete.csv")

NMDS_df <- readRDS("NMDS_df.RDS")
envfit_res <- readRDS("envfit_res.RDS")
scores <- readRDS("scores.RDS")


ps_rare <- readRDS("ps_rare_w_kmer_tree.rds") 

pair_wise_photo <- readRDS("pair_wise_photo.RDS")

post_MED_abs <- readRDS("post_MED_abs.RDS")

differential_ASV <- readRDS("differential_ASV.rds")

differential_MED <- readRDS("differential_MED.rds")
```


# 1. DADA2

## Set up the directory of each run 
Turn path on and off using as each run is completed

```{r}
path <- "../Rarotonga_Raw/"

list.files(path) # Check before moving onwards that you have the right path established
```

## Check for primers

```{r}
# Forward and reverse fastq filenames have format: SAMPLENAME_R1_001.fastq and SAMPLENAME_R2_001.fastq
fnFs <- sort(list.files(path, pattern = "R1_001.fastq.gz", full.names = TRUE))
fnRs <- sort(list.files(path, pattern = "R2_001.fastq.gz", full.names = TRUE))

fnFs
fnRs
```

```{r}
# Define Symbiodinium ITS2 primers
FWD <- "GTGAATTGCAGAACTCCGTG"
REV <- "CCTCCGCTTACTTATATGCTT"
```

```{r}
allOrients <- function(primer) {
    # Create all orientations of the input sequence
    require(Biostrings)
    dna <- DNAString(primer)  # The Biostrings works w/ DNAString objects rather than character vectors
    orients <- c(Forward = dna, Complement = Biostrings::complement(dna), Reverse = Biostrings::reverse(dna),
        RevComp = Biostrings::reverseComplement(dna))
    return(sapply(orients, toString))  # Convert back to character vector
}
FWD.orients <- allOrients(FWD)
REV.orients <- allOrients(REV)
FWD.orients
REV.orients
```

```{r}
primerHits <- function(primer, fn) {
    # Counts number of reads in which the primer is found
    nhits <- vcountPattern(primer, sread(readFastq(fn)), fixed = FALSE)
    return(sum(nhits > 0))
}

rbind(FWD.ForwardReads = sapply(FWD.orients, primerHits, fn = fnFs[[1]]),
    FWD.ReverseReads = sapply(FWD.orients, primerHits, fn = fnRs[[1]]),
    REV.ForwardReads = sapply(REV.orients, primerHits, fn = fnFs[[1]]),
    REV.ReverseReads = sapply(REV.orients, primerHits, fn = fnRs[[1]]))

# After running the primer search function above on the demultiplexed files we see that all the primers are still on the forwards and reverse reads. Use cutadapt to remove these.
```

## Cutadapt bash script

Bash script to remove primers in all orientations:

```{}
bash cutadapt_trim_primers_b.bash
```

This will launch the looped code below:

```{}
for i in *_R1_001.fastq.gz
do
  SAMPLE=$(echo ${i} | sed "s/_R1_001\.fastq\.gz//") 
  echo ${SAMPLE}_R1_001.fastq.gz ${SAMPLE}_R2_001.fastq.gz
cutadapt -b file:primers.fasta -B file:primers.fasta -n 2 -m 50 -o ${SAMPLE}.R1.CUT.fastq.gz -p ${SAMPLE}.R2.CUT.fastq.gz ${SAMPLE}_R1_001.fastq.gz ${SAMPLE}_R2_001.fastq.gz
done
```

```{r}
# List new directory of 'Cuts'. These are the files generated by the bash script above that have .CUT. in the filename.
path <- "../Rarotonga_Raw/Cuts/"
list.files(path)
```

```{r}
# Pattern search again for R1.CUT and R2.CUT in filenames
cutFs <- sort(list.files(path, pattern = "R1.CUT.fastq", full.names = TRUE))
cutRs <- sort(list.files(path, pattern = "R2.CUT.fastq", full.names = TRUE))
```

```{r}
# Check how well cutadapt worked at removing primers

rbind(FWD.ForwardReads = sapply(FWD.orients, primerHits, fn = cutFs[[1]]), 
    FWD.ReverseReads = sapply(FWD.orients, primerHits, fn = cutRs[[1]]), 
    REV.ForwardReads = sapply(REV.orients, primerHits, fn = cutFs[[1]]), 
    REV.ReverseReads = sapply(REV.orients, primerHits, fn = cutRs[[1]]))

## All primers removed!
```

```{r}
# Two rounds of strsplit to remove head and tail of the file name and keep the sample name.
sample.names <- sapply(strsplit(basename(cutFs), ".R1"), `[`, 1)
sample.names <- sapply(strsplit(sample.names, "_"), `head`, 1)
```

```{r}
# Check quality of sequences of Forward reads
plotQualityProfile(cutFs[1:3])
```

```{r}
# Check quality of sequences of Reverse reads
plotQualityProfile(cutRs[1:3])
```

```{r}
# Set up subdirectory for placing the filtered sequences
filtFs <- file.path(path, "filtered", paste0(sample.names, "_R1_filt.fastq.gz"))
filtRs <- file.path(path, "filtered", paste0(sample.names, "_R2_filt.fastq.gz"))
names(filtFs) <- sample.names
names(filtRs) <- sample.names
```

```{r}
# Filter the cutFs and cutRs into the directories created above. maxN = 0 because dada2 cannot have sequences with ambiguous bases. maxEE = c(2, 4): relaxed the EE to 4 for the reverse reads which typically are of lower quality.

out <- filterAndTrim(cutFs, filtFs, cutRs, filtRs, maxN = 0, maxEE = c(2, 4), truncQ = 2, minLen = 50, rm.phix = TRUE, compress = TRUE, multithread = FALSE) # On Windows set multithread=FALSE

# Check how many reads make it through filtering
head(out)
```

```{r}
# Create an error model for the forward reads
errF <- learnErrors(filtFs, multithread = FALSE)
```

```{r}
# Create an error model for the reverse reads
errR <- learnErrors(filtRs, multithread = FALSE)
```

```{r}
# Check that the error model fits the errors
plotErrors(errF, nominalQ = TRUE)
```

## Denoise and infer samples

```{r}
# Run the core dada2 method to remove noise and infer samples on forward reads
dadaFs <- dada(filtFs, err = errF, multithread = FALSE)
```

```{r}
# Run the core dada2 method to remove noise and infer samples on reverse reads
dadaRs <- dada(filtRs, err = errR, multithread = FALSE)
```

```{r}
# Merge the forward and reverse reads
mergers <- mergePairs(dadaFs, filtFs, dadaRs, filtRs, verbose = TRUE)
```

```{r}
# Turn the successfully merged sequences into a sequence table (i.e. an 'otu table' or 'asv table')
seqtab <- makeSequenceTable(mergers)
```

## Save point

```{r}
saveRDS(seqtab, "seqtab.rds") # Complete!
```

# 4. Remove chimeras after run merging

```{r}
seqtab.nochim <- removeBimeraDenovo(seqtab, method = "consensus", multithread = FALSE)
saveRDS(seqtab.nochim, "seqtab.nochim.rds")

sum(seqtab.nochim) / sum(seqtab) # = 0.9065372
```

# 5. Assign taxonomy:

## Assign taxonomy to reads

```{r}
seqtab.nochim <- readRDS("seqtab.nochim.rds")

tax_bs <- assignTaxonomy(seqtab.nochim, "ITS2dbn1_Dinophy_Symbio.fasta", multithread = FALSE, outputBootstraps = TRUE, minBoot = 0)
saveRDS(tax_bs, "seqtab.nochim.tax_bs.rds")
```

## Inspect bootstrap values

```{r}
tax_bs <- readRDS("seqtab.nochim.tax_bs.rds")
taxa <- as.data.frame(tax_bs$tax, stringsAsFactors = FALSE)
taxa <- tibble::rownames_to_column(taxa, var = "seq")

boot <- as.data.frame(tax_bs$boot)
boot <- tibble::rownames_to_column(boot, var = "seq")

# Inspect boostrap values

taxfilt <- left_join(taxa, boot, by = "seq", suffix = c("tax", "boot"))

# 50% bs support for symbiodiniaceae at family looks ok
# Below this it falls off to other dinoflagellates in NCBI

taxfilt <- left_join(taxa, boot, by = "seq", suffix = c("tax", "boot")) %>%
  mutate(Genustax = case_when(Familyboot < 50 & str_detect(Familytax, "Symbiodiniaceae") ~ "Unassigned",
                            TRUE ~ as.character(.$Genustax)),
         Speciestax = case_when(Familyboot < 50 & str_detect(Familytax, "Symbiodiniaceae") ~ "Unassigned",
                            TRUE ~ as.character(.$Speciestax)),
         Familytax = case_when(Familyboot < 50 & str_detect(Familytax, "Symbiodiniaceae") ~ "Unassigned",
                            TRUE ~ as.character(.$Familytax))) %>%
  mutate(Phylumtax = case_when(Phylumboot < 50  & !str_detect(Familytax, "Symbiodiniaceae") ~ "Unassigned",
                               TRUE ~ as.character(.$Phylumtax)),
         Classtax = case_when(Classboot < 50 & !str_detect(Familytax, "Symbiodiniaceae") ~ "Unassigned",
                              TRUE ~ as.character(.$Classtax)),
         Ordertax = case_when(Orderboot < 50 & !str_detect(Familytax, "Symbiodiniaceae") ~ "Unassigned",
                              TRUE ~ as.character(.$Ordertax)),
         Familytax = case_when(Familyboot < 50 & !str_detect(Familytax, "Symbiodiniaceae") ~ "Unassigned",
                               TRUE ~ as.character(.$Familytax)),
         Genustax = case_when(Genusboot < 50 & !str_detect(Familytax, "Symbiodiniaceae") ~ "Unassigned",
                              TRUE ~ as.character(.$Genustax)),
         Speciestax = case_when(Speciesboot < 1 & !str_detect(Familytax, "Symbiodiniaceae") ~ "Unassigned",
                                TRUE ~ as.character(.$Speciestax)))

rownames(taxfilt) <- taxfilt$seq
taxfilt$seq = NULL

saveRDS(taxfilt, "seqtab.nochim.tax.rds")
```

# 6. Hand-off to phyloseq



```{r}
seqtab.nochim <- readRDS("seqtab.nochim.rds")

tax <- readRDS("seqtab.nochim.tax.rds") %>%
  select(Kingdomtax, Phylumtax, Classtax, Ordertax, Familytax, Genustax, Speciestax)
taxtab <- tax_table(tax)
rownames(taxtab) <- rownames(tax)
colnames(taxtab) <- c("Kingdom", "Phylum", "Class", "Order", "Family", "Genus", "Species")

samptab <- read_csv("metadata_complete.csv")
sample_names <- samptab$sample_ID
samptab <- sample_data(samptab)
sample_names(samptab) <- sample_names

ps <- phyloseq(otu_table(seqtab.nochim, taxa_are_rows = FALSE), 
               sample_data(samptab),
               taxtab)

dna <- Biostrings::DNAStringSet(taxa_names(ps))
names(dna) <- taxa_names(ps)
ps <- merge_phyloseq(ps, dna)
taxa_names(ps) <- paste0("ASV", seq(ntaxa(ps)))

ps <- ps %>%
  subset_samples(pair != 96 & pair != 49) %>%
  subset_taxa(Family == "Symbiodiniaceae")
  

saveRDS(ps, "ps.RDS")

```

#Kmer tree, UNIFRAC
##Kmer tree
```{r}
ps_sub <- ps %>%
  subset_samples(project == "c_v_e")

ps_sub <- prune_taxa(taxa_sums(ps_sub) > 0, ps_sub)

ASVs <- refseq(ps_sub) %>%
    DNAStringSet_to_df()
  
kdist <- ASVs %>%
  df_to_DNAStringset() %>%
  DNAStringSet_to_DNAbin() %>%
  kmer::kdistance(k = 7, residues = "DNA", method = "edgar") %>%
  as.matrix()

tree <- phangorn::upgma(kdist)
write.tree(tree, file = "ITS2_kmer.tree")

# add the tree back into the phyloseq object
tree <- read_tree("ITS2_kmer.tree")
phy_tree(ps_sub) <- phy_tree(tree)

saveRDS(ps_sub, "ps_sub_w_kmer_tree.rds")
```

#pair_wise photobiology vs UniFrac

```{r}
#filter for only species with significantly differentially abundant sequnece variants
ps_hetero_ITS2 <- ps_sub %>%
  subset_samples(species == "Acropora cf anthocercis" | species == "Acropora elseyi" | species == "Acropora hyacinthus" | species == "Hydnophora microconos" | species == "Porites lutea cf. lobata" | species == "Acropora nasuta" | species == "Favia stelligera")

# Transform abundances

ps.hell <- microbiome::transform(ps_hetero_ITS2, transform = "hellinger", target = "OTU", shift = 1, scale = 1)

#Create distance matrix
ps.uni_c_v_e <- phyloseq::distance(ps.hell, method = "wunifrac") %>% as.matrix()  

long_dist <- reshape2::melt(ps.uni_c_v_e)[melt(lower.tri(ps.uni_c_v_e))$value,] %>%
  #mutate(value = (1-value)*100) %>%
  select(from = Var1, to = Var2, uni = value) %>%
  mutate(pair1 = str_sub(from, start = 2),
         pair2 = str_sub(to, start = 2)) %>%
  filter(pair1 == pair2) %>%
  mutate(pair1 = as.numeric(pair1), pair2 = as.numeric(pair2))

# Get photobiology

pair_wise_total_brightness <- left_join((total_brightness %>% rename(Sample = sampleID)), (melt %>% select(Sample, pair) %>% distinct()), by = "Sample") %>% 
  filter(Sample != "C49" & Sample != "E49" & Sample != "C96" & Sample != "E96") %>%
  filter(species == "Acropora cf anthocercis" | species == "Acropora elseyi" | species == "Acropora hyacinthus" | species == "Hydnophora microconos" | species == "Porites lutea cf. lobata" | species == "Acropora nasuta" | species == "Favia stelligera") %>%
  select(total_reflectance, pair, habitat) %>% pivot_wider(values_from = total_reflectance, names_from = habitat) %>% 
  rename(total_reflectance_cryptic = cryptic, total_reflectance_exposed = exposed)

pair_wise_photo <- pair_wise_photo %>%
  mutate(pair = as.numeric(pair)) %>%
  filter(pair != 96 & pair != 49) %>%
  filter(species == "Acropora cf anthocercis" | species == "Acropora elseyi" | species == "Acropora hyacinthus" | species == "Hydnophora microconos" | species == "Porites lutea cf. lobata" | species == "Acropora nasuta" | species == "Favia stelligera")

pair_wise_photo <- left_join(pair_wise_photo, pair_wise_total_brightness, by = "pair")

long_dist_photo <- left_join(long_dist, pair_wise_photo, by = c("pair1" = "pair")) %>%
  mutate(delta_onemQ = onemQ_cryptic - onemQ_exposed, 
         delta_onemC = onemC_exposed - onemC_cryptic, 
         delta_total_reflectance = total_reflectance_exposed - total_reflectance_cryptic) 


ggplot(long_dist_photo, aes(uni, delta_onemQ)) +
  geom_point(aes(col = species)) +
  geom_encircle(aes(fill = species), s_shape = 1, expand = 0, alpha = 0.2, show.legend = FALSE) +
  facet_wrap(~species)

ggplot(long_dist_photo, aes(uni, delta_onemC)) +
  geom_point(aes(col = species)) +
  geom_encircle(aes(fill = species), s_shape = 1, expand = 0, alpha = 0.2, show.legend = FALSE) +
  facet_wrap(~species)

ggplot(long_dist_photo, aes(uni, delta_total_reflectance)) +
  geom_point(aes(col = species)) +
  geom_encircle(aes(fill = species), s_shape = 1, expand = 0, alpha = 0.2, show.legend = FALSE) +
  facet_wrap(~species)

pair_wise_photo_Ek_FqFmmax <- FqFmparams %>%
  ungroup() %>%
  mutate(pair = str_sub(sampleID, start = 2)) %>%
  select(pair, species, term, estimate, habitat) %>%
  pivot_wider(names_from = term, values_from = estimate) %>%
  pivot_wider(names_from = habitat, values_from = FqFmmax:Ek) %>%
  mutate(pair = as.numeric(pair)) %>%
  filter(pair != 96 & pair != 49)
  
saveRDS(pair_wise_photo_Ek_FqFmmax, "pair_wise_photo_Ek_FqFmmax.RDS") 


long_dist_photo_Ek_FqFmmax <- left_join(long_dist, pair_wise_photo_Ek_FqFmmax, by = c("pair1" = "pair") ) %>% 
  mutate(delta_FqFmmax = FqFmmax_cryptic - FqFmmax_exposed, 
         delta_Ek = Ek_exposed - Ek_cryptic) 

ggplot(long_dist_photo_Ek_FqFmmax, aes(uni, delta_Ek)) +
  geom_point(aes(col = species)) +
  geom_encircle(aes(fill = species), s_shape = 1, expand = 0, alpha = 0.2, show.legend = FALSE) +
  facet_wrap(~species)


lm_Ek <- long_dist_photo_Ek_FqFmmax %>% group_by(species) %>% do(tidy(lm(uni ~ delta_Ek, data = .))) %>%
  filter(term == "delta_Ek")

lm_FqFmmax <- long_dist_photo_Ek_FqFmmax %>% group_by(species) %>% do(tidy(lm(uni ~ delta_FqFmmax, data = .))) %>%
  filter(term == "delta_FqFmmax") 

lm_onemC <- long_dist_photo %>% group_by(species) %>% do(tidy(lm(uni ~ delta_onemC, data = .))) %>%
  filter(term == "delta_onemC") 

lm_onemQ <- long_dist_photo %>% group_by(species) %>% do(tidy(lm(uni ~ delta_onemQ, data = .))) %>%
  filter(term == "delta_onemQ")

lm_total_reflectance <- long_dist_photo %>% group_by(species) %>% do(tidy(lm(uni ~ delta_total_reflectance, data = .))) %>%
  filter(term == "delta_total_reflectance")

lm_all <- rbind(lm_Ek, lm_FqFmmax, lm_onemC, lm_onemQ)

#adjust p-values

get_adjusted_pval <- function(df, pAdjustMethod = "BH", ...) {
    if(is.null(df$p.value)) {
        stop("p-value is required")
    } else {
        p <- df$p.value
        df$adjust.pvalue <- p.adjust(p, method = pAdjustMethod)
        df
    }
}

p_adjusted_all <- get_adjusted_pval(lm_all, pAdjustMethod = "BH") %>%
  filter(adjust.pvalue < 0.05)


```

## pair-wise for rare bioshpere

```{r}
#filter samples for only rare (< 0.01 freq) ASVs

rare_OTU_df <- melt %>%
  select(OTU, Sample, Abundance, habitat, species) %>%
  group_by(Sample) %>%
  mutate(rel_freq = Abundance / sum(Abundance)) %>%
  filter(rel_freq > 0, rel_freq < 0.01) %>%
  ungroup()

sample_list <- raro_metadata %>% filter(project == "c_v_e", pair != 96 & pair != 49, sample_ID != "C1") %>% select(sample_ID)
sample_list <- as.character(sample_list$sample_ID)

ps_rare <- ps %>% subset_samples(sample_ID == "C1")
rare_OTU_C1 <- rare_OTU_df %>% filter(Sample == "C1")
ps_rare <- prune_taxa(rare_OTU_C1$OTU, ps_rare)

for(i in 1:length(sample_list)) {
  
  ps_i <- subset_samples(ps, ps@sam_data$sample_ID %in% sample_list[i])
  rare_OTU_i <- rare_OTU_df %>% filter(Sample %in% sample_list[i])
  ps_i <- prune_taxa(rare_OTU_i$OTU, ps_i)
  ps_rare <- merge_phyloseq(ps_rare, ps_i)
}

ps_rare <- prune_taxa(taxa_sums(ps_rare) > 0, ps_rare)

ASVs_rare <- refseq(ps_rare) %>%
    DNAStringSet_to_df()
  
kdist_rare <- ASVs_rare %>%
  df_to_DNAStringset() %>%
  DNAStringSet_to_DNAbin() %>%
  kmer::kdistance(k = 7, residues = "DNA", method = "edgar") %>%
  as.matrix()

tree <- phangorn::upgma(kdist_rare)
write.tree(tree, file = "ITS2_kmer_rare.tree")

# add the tree back into the phyloseq object

tree_rare <- read_tree("ITS2_kmer_rare.tree")

phy_tree(ps_rare) <- phy_tree(tree_rare)

saveRDS(ps_rare, "ps_rare_w_kmer_tree.rds")
```


```{r}

#Transform abundances
ps.hell_rare <- microbiome::transform(ps_rare, transform = "hellinger", target = "OTU", shift = 1, scale = 1)

#Create distance matrix
ps_rare_dist <- phyloseq::distance(ps.hell_rare, method = "wunifrac") %>% as.matrix()  

long_dist_rare <- reshape2::melt(ps_rare_dist)[melt(lower.tri(ps_rare_dist))$value,] %>%
  #mutate(value = (1-value)*100) %>%
  select(from = Var1, to = Var2, uni = value) %>%
  mutate(pair1 = str_sub(from, start = 2),
         pair2 = str_sub(to, start = 2)) %>%
  filter(pair1 == pair2) %>%
  mutate(pair1 = as.numeric(pair1), pair2 = as.numeric(pair2))

# Get photobiology

pair_wise_total_brightness <- left_join((total_brightness %>% rename(Sample = sampleID)), (melt %>% select(Sample, pair) %>% distinct()), by = "Sample") %>% 
  filter(Sample != "C49" & Sample != "E49" & Sample != "C96" & Sample != "E96") %>%
  select(total_reflectance, pair, habitat) %>% pivot_wider(values_from = total_reflectance, names_from = habitat) %>% rename(total_reflectance_cryptic = cryptic, total_reflectance_exposed = exposed)

pair_wise_photo <- readRDS("pair_wise_photo.RDS") %>% 
   mutate(pair = as.numeric(pair)) %>%
  filter(pair != 96 & pair != 49)

pair_wise_photo <- left_join(pair_wise_photo, pair_wise_total_brightness, by = "pair")

long_dist_rare_photo <- left_join(long_dist_rare, pair_wise_photo, by = c("pair1" = "pair")) %>%
  mutate(delta_onemQ = onemQ_cryptic - onemQ_exposed, 
         delta_onemC = onemC_exposed - onemC_cryptic, 
         delta_total_reflectance = total_reflectance_exposed - total_reflectance_cryptic) 


ggplot(long_dist_rare_photo, aes(uni, delta_onemQ)) +
  geom_point(aes(col = species)) +
  geom_encircle(aes(fill = species), s_shape = 1, expand = 0, alpha = 0.2, show.legend = FALSE) +
  facet_wrap(~species)

ggplot(long_dist_rare_photo, aes(uni, delta_onemC)) +
  geom_point(aes(col = species)) +
  geom_encircle(aes(fill = species), s_shape = 1, expand = 0, alpha = 0.2, show.legend = FALSE) +
  facet_wrap(~species) 

ggplot(long_dist_rare_photo, aes(uni, delta_total_reflectance)) +
  geom_point(aes(col = species)) +
  geom_encircle(aes(fill = species), s_shape = 1, expand = 0, alpha = 0.2, show.legend = FALSE) +
  facet_wrap(~species)

pair_wise_photo_Ek_FqFmmax <- FqFmparams %>%
  ungroup() %>%
  mutate(pair = str_sub(sampleID, start = 2)) %>%
  select(pair, species, term, estimate, habitat) %>%
  pivot_wider(names_from = term, values_from = estimate) %>%
  pivot_wider(names_from = habitat, values_from = FqFmmax:Ek) %>%
  mutate(pair = as.numeric(pair))
  
saveRDS(pair_wise_photo_Ek_FqFmmax, "pair_wise_photo_Ek_FqFmmax.RDS") 


long_dist_rare_photo_Ek_FqFmmax <- left_join(long_dist_rare, pair_wise_photo_Ek_FqFmmax, by = c("pair1" = "pair") ) %>% 
  mutate(delta_FqFmmax = FqFmmax_cryptic - FqFmmax_exposed, 
         delta_Ek = Ek_exposed - Ek_cryptic) 

ggplot(long_dist_rare_photo_Ek_FqFmmax, aes(uni, delta_Ek)) +
  geom_point() +
  geom_encircle(s_shape = 1, expand = 0, alpha = 0.2, show.legend = FALSE) +
  facet_wrap(~species)

rare_lm_Ek <- long_dist_rare_photo_Ek_FqFmmax %>% group_by(species) %>% do(tidy(lm(uni ~ delta_Ek, data = .))) %>%
  filter(term == "delta_Ek")

rare_lm_FqFmmax <- long_dist_rare_photo_Ek_FqFmmax %>% group_by(species) %>% do(tidy(lm(uni ~ delta_FqFmmax, data = .))) %>%
  filter(term == "delta_FqFmmax") 

rare_lm_onemC <- long_dist_rare_photo %>% group_by(species) %>% do(tidy(lm(uni ~ delta_onemC, data = .))) %>%
  filter(term == "delta_onemC") 

rare_lm_onemQ <- long_dist_rare_photo %>% group_by(species) %>% do(tidy(lm(uni ~ delta_onemQ, data = .))) %>%
  filter(term == "delta_onemQ")

rare_lm_total_reflectance <- long_dist_rare_photo %>% group_by(species) %>% do(tidy(lm(uni ~ delta_total_reflectance, data = .))) %>%
  filter(term == "delta_total_reflectance")

rare_lm_all <- rbind(rare_lm_Ek, rare_lm_FqFmmax, rare_lm_onemC, rare_lm_onemQ)

#adjust p-values

get_adjusted_pval <- function(df, pAdjustMethod = "BH", ...) {
    if(is.null(df$p.value)) {
        stop("p-value is required")
    } else {
        p <- df$p.value
        df$adjust.pvalue <- p.adjust(p, method = pAdjustMethod)
        df
    }
}

p_adjusted_all <- get_adjusted_pval(rare_lm_all, pAdjustMethod = "BH") %>%
  filter(adjust.pvalue < 0.05)

p_adjusted_Ek <- get_adjusted_pval(rare_lm_Ek, pAdjustMethod = "BH")
p_adjusted_FqFmmax <- get_adjusted_pval(rare_lm_FqFmmax, pAdjustMethod = "BH")
p_adjusted_onemC <- get_adjusted_pval(rare_lm_onemC, pAdjustMethod = "BH")
p_adjusted_onemQ <- get_adjusted_pval(rare_lm_onemQ, pAdjustMethod = "BH")
p_adjusted_total_reflectance <- get_adjusted_pval(rare_lm_total_reflectance, pAdjustMethod = "BH")



p_adjust_df <- rbind(p_adjusted_Ek, p_adjusted_FqFmmax, p_adjusted_onemC, p_adjusted_onemQ, p_adjusted_total_reflectance) %>%
  filter(adjust.pvalue < 0.05)
```

###Hydonophora plots
```{r}
H.micro_delta_onemQ <- long_dist_rare_photo %>% 
  filter(species =="Hydnophora microconos") %>%
  ggplot(aes(uni, delta_onemQ)) +
  geom_point(aes(size = 3)) +
  geom_smooth(method='lm', formula= y~x, col = "black", alpha = 0.2) +
  ylab("Delta [1 - Q]") +
  xlab("UniFrac distance") + 
  theme_bw() + 
  theme(panel.border = element_blank(), panel.grid.major = element_blank(),
panel.grid.minor = element_blank(), axis.line = element_line(colour = "black")) +
  theme(legend.position = "none")

H.micro_delta_onemC <- long_dist_rare_photo %>% 
  filter(species =="Hydnophora microconos") %>%
  ggplot(aes(uni, delta_onemC)) +
  geom_point() +
  geom_encircle(aes(fill = species), s_shape = 1, expand = 0, alpha = 0.2, show.legend = FALSE) +
  theme(aspect.ratio = 1, axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.1)) +
  ylab("Delta [1 - C]") +
  xlab("UniFrac distance") + 
  theme_bw() + 
  theme(panel.border = element_blank(), panel.grid.major = element_blank(),
panel.grid.minor = element_blank(), axis.line = element_line(colour = "black"))

H.micro_delta_Ek <- long_dist_rare_photo_Ek_FqFmmax %>%
  filter(species =="Hydnophora microconos") %>%
  ggplot(aes(uni, delta_Ek)) +
  geom_point() +
  geom_encircle(s_shape = 1, expand = 0, alpha = 0.2, show.legend = FALSE) +
  theme(aspect.ratio = 1, axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.1)) +
  ylab("Delta Ek") +
  xlab("UniFrac distance") + 
  theme_bw() + 
  theme(panel.border = element_blank(), panel.grid.major = element_blank(), 
        panel.grid.minor = element_blank(), axis.line = element_line(colour = "black")) 

H.micro_FqFmmax <- long_dist_rare_photo_Ek_FqFmmax %>%
  filter(species =="Hydnophora microconos") %>%
  ggplot(aes(uni, delta_FqFmmax)) +
  geom_point() +
  geom_encircle(s_shape = 1, expand = 0, alpha = 0.2, show.legend = FALSE) +
  theme(aspect.ratio = 1, axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.1)) +
  ylab("Delta Fq/Fm max") +
  xlab("UniFrac distance") + 
  theme_bw() + 
  theme(panel.border = element_blank(), panel.grid.major = element_blank(), 
        panel.grid.minor = element_blank(), axis.line = element_line(colour = "black"))

H.micro_plot <- (H.micro_FqFmmax + H.micro_delta_Ek) / (H.micro_delta_onemC + H.micro_delta_onemQ)

H.micro_plot

ggsave(plot = H.micro_delta_onemQ, "H.micro_delta_onemQ.pdf", height = 5, width = 8)
```

#Plot phyloseq melt

```{r}
melt <- psmelt(ps) %>%
  mutate(Species = as.character(Species)) %>%
  mutate(Species = case_when(str_detect(Species, "|") ~ word(as.character(Species), 1, sep = "\\|"),
         TRUE ~ Species)) %>%
  mutate(Species = as.factor(Species)) %>%
  filter(Family == "Symbiodiniaceae", project == "c_v_e", Abundance > 0)

saveRDS(melt, "melt.RDS")

melt_4_deseq2 <- psmelt(ps) %>%
  mutate(Species = as.character(Species)) %>%
  mutate(Species = case_when(str_detect(Species, "|") ~ word(as.character(Species), 1, sep = "\\|"),
         TRUE ~ Species)) %>%
  mutate(Species = as.factor(Species)) %>%
  filter(Family == "Symbiodiniaceae", project == "c_v_e")

saveRDS(melt_4_deseq2, "melt_4_deseq2.RDS")

explorer <- melt %>% 
  summarise(sum(Abundance), sum(n_distinct(OTU)))

explorer <- melt %>%
  group_by(Sample) %>%
  summarise(total_reads = sum(Abundance)) %>%
  ungroup() %>%
  summarise(mean_reads = mean(total_reads), sd = sd(total_reads))

explorer <- melt %>%
  group_by(Genus) %>%
  summarise(total_reads = sum(Abundance)) %>%
  mutate(proportion = (total_reads / sum(total_reads)))
```

#Species accumilation curves
```{r}
exposed_only <- melt %>%
  filter(habitat == "exposed", Abundance > 0) %>%
  group_by(Sample) %>%
  summarise(n_distinct_ASV = n_distinct(OTU)) %>%
  ungroup() %>%
  summarise(mean_n_distinct_ASV = mean(n_distinct_ASV), sd = sd(n_distinct_ASV))

exposed_and_shaded <- melt %>%
  filter(Abundance > 0) %>%
  group_by(pair) %>%
  summarise(n_distinct_ASV = n_distinct(OTU)) %>%
  ungroup() %>%
  summarise(mean_n_distinct_ASV = mean(n_distinct_ASV), sd = sd(n_distinct_ASV))
  
exposed_only_specaccu <- melt %>%
  filter(habitat == "exposed", Abundance > 0) %>%
  select(Sample, OTU, Abundance) %>%
  pivot_wider(names_from = OTU, values_from = Abundance) %>%
  mutate_all(~replace_na(., 0)) %>%
  tibble::column_to_rownames(var = "Sample")
  
specaccum_E <- specaccum(exposed_only_specaccu, 'random')

shaded_exposed_specaccu <- melt %>%
  filter(Abundance > 0) %>%
  select(pair, OTU, Abundance) %>%
  group_by(pair, OTU) %>%
  mutate(sum_abundance = sum(Abundance)) %>%
  select(-Abundance) %>%
  distinct() %>%
  ungroup() %>%
  pivot_wider(names_from = OTU, values_from = sum_abundance) %>%
  mutate_all(~replace_na(., 0)) %>%
  tibble::column_to_rownames(var = "pair")
  
specaccum_SE <- specaccum(shaded_exposed_specaccu, 'random')

specaccum_E_df <- data.frame(specaccum_E$richness,specaccum_E$sites,specaccum_E$sd) %>%
  mutate(method = "Exposed_only") %>%
  dplyr::rename(Richness = specaccum_E.richness, Colony = specaccum_E.sites, sd = specaccum_E.sd)

specaccum_SE_df <- data.frame(specaccum_SE$richness,specaccum_SE$sites,specaccum_SE$sd) %>%
  mutate(method = "Shaded_and_exposed") %>%
    dplyr::rename(Richness = specaccum_SE.richness, Colony = specaccum_SE.sites, sd = specaccum_SE.sd)

specaccum_ggplot_df <- rbind(specaccum_E_df, specaccum_SE_df)


specaccum_curve <- specaccum_ggplot_df %>%
  ggplot(aes(Colony, Richness, colour = method)) +
  geom_line(size = 1.5) +
  scale_colour_manual(values=c(Exposed_only = "royalblue", Shaded_and_exposed ="red"), labels = c("Exposed only", "Pooled shaded and exposed")) +
  ylab("ITS2 sequence richness") +
  xlab("Colonies sampled") + 
  labs(colour = "Colony sampling method") +
  #scale_colour_discrete(labels = c("Combined shaded and exposed", "Exposed only")) +
  guides(colour = guide_legend(reverse=TRUE)) +
  theme_bw() + 
  theme(panel.border = element_blank(),
        panel.grid.minor = element_blank(), 
        axis.line = element_line(colour = "black"),
        axis.text = element_text(size = 12),
        axis.title = element_text(size = 16),
        legend.text=element_text(size = 14),
        legend.title = element_text(size = 16))

ggsave(plot = specaccum_curve, "specaccum_curve.tiff", height = 5, width = 8)

```


#SymPortal post_MED
```{r}

post_MED_abs <- read.table("post_MED_seqs.absolute.abund_only.txt", header = TRUE, sep = '\t') %>%
  pivot_longer(!sample_uid, names_to = "seq_name", values_to = "Abundance")

post_MED_sample_names <- read.table("post_MED_seqs.absolute.abund_and_meta.txt", header = TRUE, sep = '\t') %>%
  select(sample_uid, sample_name) %>%
  mutate(sample_uid = as.numeric(sample_uid))

post_MED_abs <- left_join(post_MED_sample_names, post_MED_abs, by = "sample_uid") %>%
  rename(sample_ID = sample_name) %>%
  select(-sample_uid) #%>%
  #filter(Abundance > 0)

post_MED_abs <- left_join(post_MED_abs, raro_metadata, by = "sample_ID") %>%
  filter(pair != 96 & pair != 49)

saveRDS(post_MED_abs, "post_MED_abs.RDS")
```

## summary stats
```{r}
post_MED_abs %>%
  summarise(total_abundance = sum(Abundance), n_unique_seq = n_distinct(seq_name))

post_MED_abs %>%
  group_by(sample_ID) %>%
  summarise(sum_seq = sum(Abundance)) %>%
  summarise(mean(sum_seq), sd(sum_seq), min(sum_seq), max(sum_seq))

```

#SymPortal ITS2 profiles

```{r}
#absolute abundance

raro_metadata <- read.csv("metadata_complete.csv")

ITS2_profiles_abs <- read.delim("profiles.absolute.abund_and_meta.txt", skip = 6) %>% 
  rename(sample_ID = X) %>%
  select(-ITS2.type.profile)

ITS2_profiles_abs <- ITS2_profiles_abs %>%
  pivot_longer(-sample_ID, names_to = "ITS2_profile", values_to = "absolute_abundance") %>% 
  left_join(raro_metadata, ITS2_profiles_abs, by = "sample_ID") %>%
  mutate(absolute_abundance = as.numeric(absolute_abundance)) 

ITS2_profiles_abs_cve <- ITS2_profiles_abs %>%
  filter(project == "c_v_e") %>%
  filter(absolute_abundance > 0) %>%
  filter(pair != 96 & pair != 49)

#relative abundance

ITS2_profiles_rel <- read.delim("profiles.relative.abund_and_meta.txt", skip = 6) %>% 
  rename(sample_ID = X) %>%
  select(-ITS2.type.profile)

ITS2_profiles_rel <- ITS2_profiles_rel %>%
  pivot_longer(-sample_ID, names_to = "ITS2_profile", values_to = "relative_abundance") %>%
  left_join(raro_metadata, ITS2_profiles_rel, by = "sample_ID") %>%
  mutate(relative_abundance = as.numeric(relative_abundance))

ITS2_profiles_rel_cve <- ITS2_profiles_rel %>%
  filter(project == "c_v_e") %>%
  filter(relative_abundance > 0) %>%
  filter(pair != 96 & pair != 49)

ITS2_profiles_bind <- ITS2_profiles_rel_cve %>%
  group_by(sample_ID) %>%
  mutate(other_prop = 1 - sum(relative_abundance)) %>%
  ungroup() %>%
  mutate(other_name = "Other Symbiodiniaceae") %>%
  select(-relative_abundance, -ITS2_profile) %>%
  rename(relative_abundance = other_prop, ITS2_profile = other_name) %>%
  distinct()

ITS2_profiles_rel_cve <- rbind(ITS2_profiles_rel_cve, ITS2_profiles_bind) %>%
  mutate(ITS2_profile = str_replace_all(ITS2_profile, "[.]", "-")) %>%
  mutate(ITS2_profile = str_replace_all(ITS2_profile, "C3f-C3-C50a-C3fl-C3ae-C3fm-C50q-C3b", "C3f/C3-C50a-C3fl-C3ae-C3fm-C50q-C3b")) %>%
   mutate(ITS2_profile = str_replace_all(ITS2_profile, "C1-C3-C1c-C1b-C72k-C1w", "C1/C3-C1c-C1b-C72k-C1w")) %>%
   mutate(ITS2_profile = str_replace_all(ITS2_profile, "C3f-C3fl-C3ae-C3-C3bj-C50a-C3h", "C3f/C3fl-C3ae-C3-C3bj-C50a-C3h")) %>%
   mutate(ITS2_profile = str_replace_all(ITS2_profile, "C1-C1b-C3-C1u-C1bo", "C1/C1b/C3-C1u-C1bo")) %>%
   mutate(ITS2_profile = str_replace_all(ITS2_profile, "C21-C3", "C21/C3")) %>%
   mutate(ITS2_profile = str_replace_all(ITS2_profile, "C42a-C1b-C1-C42.2-C42b", "C42a/C1b/C1-C42.2-C42b")) %>%
   mutate(ITS2_profile = str_replace_all(ITS2_profile, "C42g-C42a-C1-C42.2-C42h-C42b-C1b", "C42g/C42a/C1/C42.2-C42h-C42b-C1b")) %>%
   mutate(ITS2_profile = str_replace_all(ITS2_profile, "C42.2-C1-C42i-C3-C1b-C1au", "C42.2/C1-C42i-C3-C1b-C1au")) %>%
   mutate(ITS2_profile = str_replace_all(ITS2_profile, "C1-C1c-C1b", "C1/C1c-C1b")) %>%
   mutate(ITS2_profile = str_replace_all(ITS2_profile, "C1-C42.2-C3-C1b", "C1/C42.2/C3-C1b")) %>%
   mutate(ITS2_profile = str_replace_all(ITS2_profile, "D1-D4-D1d-D6", "D1/D4/D1d-D6")) %>%
   mutate(ITS2_profile = str_replace_all(ITS2_profile, "D4-D1-D1ab-D6", "D4/D1/D1ab-D6")) 

```

## summary stats
```{r}
summarise(ITS2_profiles_abs_cve, n_distinct(ITS2_profile))

ITS2_profiles_abs_cve %>%
  select(sample_ID, species, ITS2_profile, absolute_abundance) %>%
  group_by(sample_ID) %>%
  summarize(n_profiles = n_distinct(ITS2_profile)) %>%
  summarise(sum(n_profiles == 1), sum(n_profiles == 2), sum(n_profiles == 3), sum(n_profiles == 4), sum(n_profiles == 5))

ITS2_profiles_abs_cve %>%
  filter(genus == "Pocillopora") %>%
  group_by(pair) %>%
  summarise(n_profiles = n_distinct(ITS2_profile)) %>%
  summarise(sum(n_profiles == 1), sum(n_profiles == 2), sum(n_profiles == 3), sum(n_profiles == 4), sum(n_profiles == 5), sum(n_profiles == 6), sum(n_profiles == 7), mean(n_profiles), sd(n_profiles))


ITS2_profiles_rel_cve %>%
  filter(ITS2_profile != "Other Symbiodiniaceae") %>%
  group_by(sample_ID) %>%
  summarise(ITS2_sum = sum(relative_abundance)) %>%
  ungroup() %>%
  summarise(mean(ITS2_sum), sd(ITS2_sum))
```

# PERMANOVA ITS2 profiles
```{r}
ITS2_profiles_bray_curtis <- ITS2_profiles_abs_cve %>%
  select(sample_ID, ITS2_profile, absolute_abundance) %>%
  pivot_wider(names_from = ITS2_profile, values_from = absolute_abundance) %>%
  tibble::column_to_rownames(var = "sample_ID") %>%
  replace(is.na(.), 0) %>%
  vegdist(method = "bray")

sample_df <- ITS2_profiles_abs_cve %>%
  select(sample_ID, species, habitat) %>% 
  distinct() %>%
  tibble::column_to_rownames(var = "sample_ID")

adonis(ITS2_profiles_bray_curtis ~ species * habitat, data = sample_df)

# Terms added sequentially (first to last)
# 
#                  Df SumsOfSqs MeanSqs F.Model      R2 Pr(>F)    
# species          19    44.949 2.36574  9.1629 0.51321  0.001 ***
# habitat           1     0.292 0.29245  1.1327 0.00334  0.295    
# species:habitat  19     2.066 0.10873  0.4211 0.02359  1.000    
# Residuals       156    40.277 0.25819         0.45987           
# Total           195    87.585                 1.00000           
# ---


```


##bar plots
```{r}
Profile_list <- ITS2_profiles_rel_cve$ITS2_profile %>%
  unique()
Profile_list

#brewer.pal(6, "Blues")
C1_pal <- c("navyblue", "blue2", "steelblue2", "slateblue3", "lightseagreen", "cadetblue1")
names(C1_pal) <- c("C1/C3-C1c-C1b-C72k-C1w", "C1/C42.2/C3-C1b", "C1-C1b-C1c-C42.2-C1br-C1bh-C1cb-C72k", "C1/C1b/C3-C1u-C1bo", "C1/C1c-C1b", "C1y")

#brewer.pal(8, "Greens")
C3_pal <- c("green4", "limegreen", "olivedrab3", "springgreen", "mediumseagreen", "mediumspringgreen", "darkseagreen4", "darkseagreen")
names(C3_pal) <- c("C3f/C3-C50a-C3fl-C3ae-C3fm-C50q-C3b", "C3-C1", "C3f-C3fl-C3-C50a-C3ae-C3fm-C3h-C50q","C3k", "C3f/C3fl-C3ae-C3-C3bj-C50a-C3h", "C3fn", "C3f-C3fl-C50a-C3-C3ae-C3fm-C3h-C3bj", "C3-C3fo-C3a")

C15_pal <- c("violet", "violetred")
names(C15_pal) <- c("C15-C15ek-C15gy-C15gk-C15el", "C15-C15by")

#brewer.pal(6, "Oranges")
C42_pal <- c("yellow4", "yellow2", "orange3", "orange", "khaki1", "darkgoldenrod1")
names(C42_pal) <- c("C42.2/C1-C42i-C3-C1b-C1au", "C42.2-C42aq-C1-C42ar", "C42a/C1b/C1-C42.2-C42b", "C42g/C42a/C1/C42.2-C42h-C42b-C1b", "C42a-C42.2-C1-C1b-C42b-C1au", "C42a-C1-C42.2-C1j-C1b-C42as-C42aa")

#brewer.pal(5, "Reds")
D1_pal <- c("darkred", "red", "orangered", "firebrick", "coral") 
names(D1_pal) <- c("D4/D1/D1ab-D6", "D1-D4-D4c-D1c-D2", "D1-D4-D4c-D6-D1d-D2", "D1-D4-D6-D1b-D1d-D6c-D4d", "D1/D4/D1d-D6")

A1_pal <- c("chocolate", "chocolate4")
names(A1_pal) <- c("A1-A1ee", "A1-A1bw-A1bf-A1bx")

B1_pal <- c("#D95F0E")
names(B1_pal) <- c("B1-B1a-B1b-B1g")

C91_pal <- c("plum4")
names(C91_pal) <- c("C91")

C21_pal <- c("peachpuff")
names(C21_pal) <- c("C21/C3")

Other_pal <- c("darkgrey")
names(Other_pal) <- c("Other Symbiodiniaceae")


all_pal <- c(C1_pal, C3_pal, C15_pal, C42_pal, D1_pal, A1_pal, B1_pal, C91_pal, C21_pal, Other_pal)

Fig_4 <- ITS2_profiles_rel_cve %>%
  filter(genus == "Acropora" | species == "Favia stelligera" | species == "Leptoria phrygia") %>%
  mutate(habitat = fct_relevel(habitat, c("exposed", "cryptic"))) %>%
  ggplot(aes(pair, relative_abundance)) +
  geom_bar(stat = "identity", aes(fill = ITS2_profile), position = "fill") +
  scale_fill_manual(values = all_pal) +
  scale_y_continuous(labels = scales::percent) +
  facet_grid(habitat ~ species, scales = "free_x") +
  theme(aspect.ratio = .5) +
  theme(legend.direction = 'vertical') +
  guides(fill = guide_legend(ncol = 4)) +
  theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5), legend.position = "left") +
  ylab("ITS2 profile relative abundance (%)") +
  xlab("Sample pair") +
  theme_bw() +
  theme(panel.border = element_blank(), panel.grid.major = element_blank(),
panel.grid.minor = element_blank(), axis.line = element_line(colour = "black")) +
  theme(strip.text.x = element_text(face = "italic")) +
  theme(legend.position = "bottom") 

Fig_4

#ggsave("ITS2_Profile_Relative_Abdundance_1.tiff", plot = Fig_4, units="in", width= 14, height=7)



Fig_5 <- ITS2_profiles_rel_cve %>%
  filter(genus != "Acropora" & species != "Favia stelligera" & species != "Leptoria phrygia" & pair != 96) %>% 
  mutate(pair = as.factor(pair)) %>% filter(pair != 49) %>% mutate(pair = fct_drop(pair)) %>% 
  mutate(habitat = fct_relevel(habitat, c("exposed", "cryptic"))) %>%
  ggplot(aes(pair, relative_abundance)) +
  geom_bar(stat = "identity", aes(fill = ITS2_profile), position = "fill") +
  scale_fill_manual(values = all_pal) +
  scale_y_continuous(labels = scales::percent) +
  facet_grid(habitat ~ species, scales = "free_x") +
  theme(aspect.ratio = .5) +
  theme(legend.direction = 'vertical') +
  guides(fill = guide_legend(ncol = 4)) +
  theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5), legend.position = "left") +
  ylab("ITS2 profile relative abundance (%)") +
  xlab("Sample pair") +
  theme_bw() +
  theme(panel.border = element_blank(), panel.grid.major = element_blank(),
panel.grid.minor = element_blank(), axis.line = element_line(colour = "black")) +
  theme(strip.text.x = element_text(face = "italic")) +
  theme(legend.position = "bottom") 

Fig_5

#ggsave("ITS2_Profile_Relative_Abdundance_2.tiff", plot = Fig_5, units="in", width= 18, height=8)


```

# Differential abundance testing (DESeq2)
##Presence/absence 
```{r}
presence_absence_melt <- melt %>%
  select(OTU, Abundance, Sample, habitat, species) %>%
  group_by(OTU, habitat) %>%
  summarise(sum_Abundance = sum(Abundance)) %>%
  pivot_wider(names_from = habitat, values_from = sum_Abundance) %>%
  mutate(niche = case_when(cryptic > 0 & exposed == 0 ~ "Cryptic_only", exposed > 0 & cryptic == 0 ~ "Exposed_only", TRUE ~ "Both")) %>%
  ungroup()
  

presence_absence_melt %>%
  summarise(sum(niche == "Cryptic_only"), sum(niche == "Exposed_only"), sum(niche == "Both"))

set1 <- paste(rep("word_" , 200) , sample(c(1:1000) , 200 , replace=F) , sep="")
set2 <- paste(rep("word_" , 200) , sample(c(1:1000) , 200 , replace=F) , sep="")

#venn diagram 

exposed_list <- melt %>%
  select(OTU, Abundance, Sample, habitat, species) %>%
  group_by(OTU, habitat) %>%
  summarise(sum_Abundance = sum(Abundance)) %>%
  filter(habitat == "exposed") %>%
  filter(!sum_Abundance == 0) %>%
  select(OTU) %>%
  ungroup() %>%
  as.list()

shaded_list <- melt %>%
  select(OTU, Abundance, Sample, habitat, species) %>%
  group_by(OTU, habitat) %>%
  summarise(sum_Abundance = sum(Abundance)) %>%
  filter(habitat == "cryptic") %>%
  filter(!sum_Abundance == 0) %>%
  select(OTU) %>%
  ungroup() %>%
  as.list()

my_col <- c("red", "royalblue")

venn.diagram(x = c(exposed_list, shaded_list),
  category.names = c("Exposed only" , "Shaded only"),
  filename = 'ASV_venn_diagram.pdf',
  fill = my_col,
  cex = 2,
  cat.cex = 2,
  cat.fontface = "bold",
  height = 4000 , 
  width = 4000,
  cat.pos = c(-20, 20))
```


## DADA2 ASVs

```{r}
gm_mean = function(x, na.rm = TRUE){exp(sum(log(x[x > 0]), na.rm = na.rm) / length(x))}

alpha <- 0.05
restax <- as.data.frame(ps@tax_table@.Data)
restax <- rownames_to_column(restax, var = "ASVID")

# Compare across niches

species_list <- c("Acropora cf anthocercis","Acropora elseyi","Acropora humilis","Acropora hyacinthus","Acropora nasuta","Acropora retusa","Acropora verweyi","Favia stelligera","Leptoria phrygia","Galaxea fascicularis","Hydnophora exesa","Hydnophora microconos","Pavona duerdeni","Psammocora stellata","Psammocora obtusangula","Leptastrea purpurea","Echinopora gemmacea","Pocillopora meandrina","Pocillopora damicornis","Porites lutea cf. lobata")

differential_abundance <- data.frame()
  
for(i in 1:length(species_list)) {

  ps_ds2 <- phyloseq_to_deseq2((ps %>% 
                                subset_samples(project == "c_v_e") %>%
                                subset_samples(species == species_list[i]) %>%
                                filter_taxa(function(x) sum(x) > 0, TRUE)), ~ habitat)
  geoMeans <- apply(counts(ps_ds2), 1, gm_mean)
  estszef <- estimateSizeFactors(ps_ds2, geoMeans = geoMeans)
  ps_ds2 <- DESeq(estszef, fitType = "parametric")
  res.w.s <- results(ps_ds2, cooksCutoff = FALSE, contrast = c("habitat", "exposed", "cryptic"), tidy = TRUE) %>%
  select(-padj) %>%
  left_join(., restax, by = c("row" = "ASVID")) %>%
  mutate(test = "exposed_vs_cryptic") %>% 
  mutate(species = species_list[i])

  differential_abundance <- rbind(res.w.s, differential_abundance)

}

#adjust pvalues for all ASV tests across all species 

differential_abundance <- differential_abundance %>%
  ungroup() %>%
  mutate(padj = p.adjust(pvalue, method = "BH")) %>%
  filter(padj < 0.05)

differential_abundance_list <- differential_abundance %>%
  select(row, baseMean, log2FoldChange, species) %>%
  rename(OTU = row)

species_list <- c("Acropora cf anthocercis","Acropora elseyi","Acropora humilis","Acropora hyacinthus","Acropora nasuta","Acropora retusa","Acropora verweyi","Favia stelligera","Leptoria phrygia","Galaxea fascicularis","Hydnophora exesa","Hydnophora microconos","Pavona duerdeni","Psammocora stellata","Psammocora obtusangula","Leptastrea purpurea","Pocillopora meandrina","Porites lutea cf. lobata")

melt_sub <- melt_4_deseq2 %>%
  select(OTU, Genus, Species, Sample, Abundance, habitat, species)

differential_ASV <- data.frame()

for(i in 1:length(species_list)) {
  
  differential_abundance_i <- differential_abundance_list %>% filter(species %in% species_list[i])
  
  melt_sub_i <- melt_sub %>% filter(OTU %in% differential_abundance_i$OTU, species %in% differential_abundance_i$species)
  
  differential_ASV_i <- left_join(differential_abundance_i, melt_sub_i, by = c("OTU", "species")) %>%
  group_by(OTU, habitat) %>%
  mutate(n_absent = sum(Abundance == 0))

  #Only keep ASVs detected in at least 3 samples in at least one habitat
  differential_ASV_i_filter <- differential_ASV_i %>%
  select(OTU, habitat, n_absent) %>%
  distinct() %>%
  pivot_wider(names_from = "habitat", values_from = "n_absent") %>%
  group_by(OTU) %>%
  filter(exposed <= 1 | cryptic <= 1)
  

differential_ASV_i_corrected <- differential_ASV_i %>%
  filter(OTU %in% differential_ASV_i_filter$OTU)

differential_ASV <- rbind(differential_ASV, differential_ASV_i_corrected)
}

saveRDS(differential_ASV, "differential_ASV.rds")
```

###plot LFCs
```{r}

differential_ASV %>%
  group_by(species) %>%
  summarise(sumASV = n_distinct(OTU))


differential_abundance_freq <- differential_ASV %>%
  group_by(species) %>%
  summarise(sum_dif_abundant = sum(Abundance))

differntial_abundance_melt_total <- melt %>%
  filter(project == "c_v_e") %>% 
  filter(Abundance > 0) %>%
  select(OTU, Abundance, species, habitat) %>%
  group_by(species) %>%
  summarise(total_Abundance = sum(Abundance))

differential_abundance_freq <- left_join(differential_abundance_freq, differntial_abundance_melt_total, by = "species")

differential_abundance_freq %>%
  mutate(freq = sum_dif_abundant/total_Abundance) %>%
  summarise(mean_rel_freq = mean(freq), sd = sd(freq))

differential_abundance_freq %>%
  filter(species != "Porites lutea cf. lobata") %>%
  mutate(freq = sum_dif_abundant/total_Abundance) %>%
  summarise(mean_rel_freq = mean(freq), sd = sd(freq))

#cyrtic or exposed or both presence

differential_ASV %>%
  select(OTU, species, log2FoldChange) %>% 
  mutate(Direction = if_else(log2FoldChange > 0, "exposed", "cryptic")) %>%
  group_by(OTU) %>%
  mutate(incidence_cryptic = sum(Direction == "cryptic")) %>%
  mutate(incidence_exposed = sum(Direction == "exposed")) %>%
  mutate(niche = case_when(incidence_cryptic > 0 & incidence_exposed == 0 ~ "Cryptic_only", incidence_exposed > 0 & incidence_cryptic == 0 ~ "Exposed_only",
                           TRUE ~ "Both")) %>%
  select(OTU, niche) %>%
  distinct()

# differential_abundance_freq %>%
#   ungroup() %>%
#  summarise(sum(niche == "Cryptic_only"), sum(niche == "Exposed_only"), sum(niche == "Both"))

ASV_order <- differential_ASV %>%
  ungroup() %>%
  select(OTU, species) %>%
  select(OTU) %>%
  distinct() %>%
  as.list

 Deseq2_pal <- c("#E69F00", "#56B4E9", "#009E73", "#F0E442", "#0072B2", "#D55E00", "#CC79A7")
 names(Deseq2_pal) <- c("Acropora cf anthocercis", "Acropora elseyi", "Acropora hyacinthus", "Hydnophora microconos", "Porites lutea cf. lobata", "Acropora nasuta", "Favia stelligera")
  
differential_ASV %>%
  ungroup() %>%
  mutate(OTU = fct_relevel(OTU, ASV_order)) %>%
  mutate(Direction = if_else(log2FoldChange > 0, "exposed", "cryptic")) %>%
  ggplot(aes(x = OTU, y = log2FoldChange, shape = Direction, fill = species)) +
  geom_point(size = 3) +
  guides(shape = guide_legend(order=1)) +
  scale_shape_manual(values = c(exposed = 22, cryptic = 21), labels = c("Exposed", "Shaded")) +
  scale_fill_manual(values = Deseq2_pal, 
                    name = "Species", 
                    guide = guide_legend(override.aes = aes(shape=21), 
                    label.theme = element_text(size = 10, face = "italic"))) +
  geom_abline(intercept = 0, slope = 0, linetype = 3) +
  theme_bw() + 
  theme(panel.border = element_rect(colour = "black", fill=NA), panel.grid.major = element_blank(), 
        panel.grid.minor = element_blank(), axis.line = element_line(colour = "black")) +
  theme(axis.text.x = element_blank()) +
  #facet_wrap(~species, scales = "free_x", ncol = 2) +
  xlab("ASV") +
  ylab("Log2 fold change") +
  theme(strip.text = element_text(face = "italic")) +
  theme(legend.box = "vertical", legend.text = element_text(size = 10))
  


ggsave("Fig_differential_abundance.pdf", units = "in", width = 7, height = 5)


```

## Symportal post-MED

```{r}
# post_MED_count_matrix <- post_MED_abs %>% 
#   filter(project == "c_v_e") %>%
#   select(seq_name, sample_ID, Abundance) %>%
#   pivot_wider(names_from = sample_ID, values_from = Abundance) %>%
#   tibble::column_to_rownames(var = "seq_name") %>%
#   replace(is.na(.), 0)
# 
# 
# post_MED_col_data <- post_MED_abs %>% 
#   filter(project == "c_v_e") %>%
#   select(sample_ID, habitat, species) %>%
#   distinct() %>%
#   mutate(habitat = as.factor(habitat), species = as.factor(species)) %>%
#   tibble::column_to_rownames(var = "sample_ID")



gm_mean = function(x, na.rm = TRUE){exp(sum(log(x[x > 0]), na.rm = na.rm) / length(x))}

alpha <- 0.05
restax <- as.data.frame(ps@tax_table@.Data)
restax <- rownames_to_column(restax, var = "ASVID")

# Compare across niches

species_list <- c("Acropora cf anthocercis","Acropora elseyi","Acropora humilis","Acropora hyacinthus","Acropora nasuta","Acropora retusa","Acropora verweyi","Favia stelligera","Leptoria phrygia","Galaxea fascicularis","Hydnophora exesa","Hydnophora microconos","Pavona duerdeni","Psammocora stellata","Psammocora obtusangula","Leptastrea purpurea","Pocillopora meandrina","Porites lutea cf. lobata")

#"Pocillopora damicornis",

differential_abundance_MED <- data.frame()
  
for(i in 1:length(species_list)) {

MED_ds2 <- DESeqDataSetFromMatrix(countData = (post_MED_abs %>% filter(project == "c_v_e") %>% filter(species == species_list[i]) %>% select(seq_name, sample_ID, Abundance) %>% pivot_wider(names_from = sample_ID, values_from = Abundance) %>% tibble::column_to_rownames(var = "seq_name") %>% replace(is.na(.), 0)), 
                                    colData = (post_MED_abs %>% filter(project == "c_v_e") %>% filter(species == species_list[i]) %>% select(sample_ID, habitat, species) %>% distinct() %>% mutate(habitat = as.factor(habitat), species = as.factor(species)) %>% tibble::column_to_rownames(var = "sample_ID")), 
                                    design = ~ habitat)
  
  geoMeans <- apply(counts(MED_ds2), 1, gm_mean)
  estszef <- estimateSizeFactors(MED_ds2, geoMeans = geoMeans)
  MED_ds2 <- DESeq(estszef, fitType = "parametric")
  res.w.s <- results(MED_ds2, cooksCutoff = FALSE, contrast = c("habitat", "exposed", "cryptic"), tidy = TRUE) %>%
  select(-padj) %>%
  left_join(., restax, by = c("row" = "ASVID")) %>%
  mutate(test = "exposed_vs_cryptic") %>% 
  mutate(species = species_list[i])

    differential_abundance_MED <- rbind(res.w.s, differential_abundance_MED)

}

#add species that fail within the loop
MED_ds2 <- DESeqDataSetFromMatrix(countData = (post_MED_abs %>% filter(project == "c_v_e") %>% filter(species == "Echinopora gemmacea") %>% select(seq_name, sample_ID, Abundance) %>% pivot_wider(names_from = sample_ID, values_from = Abundance) %>% tibble::column_to_rownames(var = "seq_name") %>% replace(is.na(.), 0)), 
                                    colData = (post_MED_abs %>% filter(project == "c_v_e") %>% filter(species == "Echinopora gemmacea") %>% select(sample_ID, habitat, species) %>% distinct() %>% mutate(habitat = as.factor(habitat), species = as.factor(species)) %>% tibble::column_to_rownames(var = "sample_ID")), 
                                    design = ~ habitat)
  
  geoMeans <- apply(counts(MED_ds2), 1, gm_mean)
  estszef <- estimateSizeFactors(MED_ds2, geoMeans = geoMeans)
  MED_ds2 <- DESeq(estszef, fitType = "parametric")
  res.w.s <- results(MED_ds2, cooksCutoff = FALSE, contrast = c("habitat", "exposed", "cryptic"), tidy = TRUE) %>%
  select(-padj) %>%
  left_join(., restax, by = c("row" = "ASVID")) %>%
  mutate(test = "exposed_vs_cryptic") %>% 
  mutate(species = "Echinopora gemmacea")

    differential_abundance_MED <- rbind(res.w.s, differential_abundance_MED)
  
MED_ds2 <- DESeqDataSetFromMatrix(countData = (post_MED_abs %>% filter(project == "c_v_e") %>% filter(species == "Pocillopora damicornis") %>% select(seq_name, sample_ID, Abundance) %>% pivot_wider(names_from = sample_ID, values_from = Abundance) %>% tibble::column_to_rownames(var = "seq_name") %>% replace(is.na(.), 0)), 
                                    colData = (post_MED_abs %>% filter(project == "c_v_e") %>% filter(species == "Pocillopora damicornis") %>% select(sample_ID, habitat, species) %>% distinct() %>% mutate(habitat = as.factor(habitat), species = as.factor(species)) %>% tibble::column_to_rownames(var = "sample_ID")), 
                                    design = ~ habitat)
  
  geoMeans <- apply(counts(MED_ds2), 1, gm_mean)
  estszef <- estimateSizeFactors(MED_ds2, geoMeans = geoMeans)
  MED_ds2 <- DESeq(estszef, fitType = "parametric")
  res.w.s <- results(MED_ds2, cooksCutoff = FALSE, contrast = c("habitat", "exposed", "cryptic"), tidy = TRUE) %>%
  select(-padj) %>%
  left_join(., restax, by = c("row" = "ASVID")) %>%
  mutate(test = "exposed_vs_cryptic") %>% 
  mutate(species = "Pocillopora damicornis")

    differential_abundance_MED <- rbind(res.w.s, differential_abundance_MED)

    
#adjust pvalues for all ASV tests across all species 

differential_abundance_MED <- differential_abundance_MED %>%
  ungroup() %>%
  mutate(padj = p.adjust(pvalue, method = "BH")) %>%
  filter(padj < 0.05)

#save df
saveRDS(differential_abundance_MED, "differential_abundance_MED.rds")

#apply filter to Deseq2 results

differential_MED_abundance_list <- differential_abundance_MED %>%
  select(row, baseMean, log2FoldChange, species) %>%
  rename(seq = row)

species_list <- c("Acropora cf anthocercis","Acropora elseyi","Acropora humilis","Acropora hyacinthus","Acropora nasuta","Acropora retusa","Acropora verweyi","Favia stelligera","Leptoria phrygia","Galaxea fascicularis","Hydnophora exesa","Hydnophora microconos","Pavona duerdeni","Psammocora stellata","Psammocora obtusangula","Leptastrea purpurea","Pocillopora meandrina","Porites lutea cf. lobata")

post_MED_sub <- post_MED_abs %>%
  select(seq_name, Abundance, sample_ID, habitat, species) %>%
  dplyr::rename(seq = seq_name, sample = sample_ID)

differential_MED <- data.frame()

for(i in 1:length(species_list)) {
  
  differential_MED_abundance_i <- differential_MED_abundance_list %>% filter(species %in% species_list[i])
  
  post_MED_sub_i <- post_MED_sub %>% filter(seq %in% differential_MED_abundance_i$seq, species %in% differential_MED_abundance_i$species)
  
  differential_MED_i <- left_join(differential_MED_abundance_i, post_MED_sub_i, by = c("seq", "species")) %>%
  group_by(seq, habitat) %>%
  mutate(n_absent = sum(Abundance == 0))

  #Only keep post-MED seqs detected in at least 3 samples in at least one habitat
  differential_MED_i_filter <- differential_MED_i %>%
  select(seq, habitat, n_absent) %>%
  distinct() %>%
  pivot_wider(names_from = "habitat", values_from = "n_absent") %>%
  group_by(seq) %>%
  filter(exposed <= 1 | cryptic <= 1)
  

differential_MED_i_corrected <- differential_MED_i %>%
  filter(seq %in% differential_MED_i_filter$seq)

differential_MED <- rbind(differential_MED, differential_MED_i_corrected)
}

saveRDS(differential_MED, "differential_MED.rds")
```

###plot LFCs
```{r}
differential_MED %>%
  group_by(species) %>%
  summarise(sum_seq = n_distinct(seq))


differential_abundance_freq <- differential_MED %>%
  group_by(species) %>%
  summarise(sum_dif_abundant = sum(Abundance))

post_MED_abs_total <- post_MED_abs %>%
  filter(project == "c_v_e") %>% 
  dplyr::rename(seq = seq_name) %>%
  filter(Abundance > 0) %>%
  select(seq, Abundance, species, habitat) %>%
  group_by(species) %>%
  summarise(total_Abundance = sum(Abundance))

differential_abundance_freq <- left_join(differential_abundance_freq, post_MED_abs_total, by = "species")

differential_abundance_freq %>%
  mutate(freq = sum_dif_abundant/total_Abundance) %>%
  summarise(mean_rel_freq = mean(freq), sd = sd(freq))

MED_order <- differential_MED %>%
  ungroup() %>%
  select(seq, species) %>%
  select(seq) %>%
  distinct() %>%
  as.list

Deseq2_pal <- c("#E69F00", "#56B4E9", "#009E73", "#F0E442", "#0072B2", "#D55E00", "#CC79A7")
names(Deseq2_pal) <- c("Acropora cf anthocercis", "Acropora elseyi", "Acropora hyacinthus", "Hydnophora microconos", "Porites lutea cf. lobata", "Acropora nasuta", "Favia stelligera")
  

differential_MED %>%
  ungroup() %>%
  rename(Species = species) %>%
  mutate(seq = fct_relevel(seq, MED_order)) %>%
  mutate(Direction = if_else(log2FoldChange > 0, "exposed", "cryptic")) %>%
  ggplot(aes(x = seq, y = log2FoldChange, shape = Direction, fill = Species)) +
  geom_point(size = 3) +
  guides(shape = guide_legend(order=1)) +
  scale_shape_manual(values = c(exposed = 22, cryptic = 21), labels = c("Exposed", "Shaded")) +
  scale_fill_manual(values = Deseq2_pal,  
                    guide = guide_legend(override.aes = aes(shape=21), label.theme = element_text(size = 10, face = "italic"))) +
  geom_abline(intercept = 0, slope = 0, linetype = 3) +
  theme_bw() + 
  theme(panel.border = element_rect(colour = "black", fill=NA), panel.grid.major = element_blank(), 
        panel.grid.minor = element_blank(), axis.line = element_line(colour = "black"), axis.text.x = element_blank()) +
  xlab("Post-MED sequence") +
  ylab("Log2 fold change") +
  theme(strip.text = element_text(face = "italic"))


ggsave("Fig_differential_abundance_MED.pdf", units = "in", width = 7, height = 5)
```


```{r}
#confirm presence of post-MED seqs as DIVs

differential_MED_list <- differential_MED %>%
  ungroup() %>%
  select(seq) %>%
  distinct()

DIV_list <- c("D1", "D4", "D6", "D1b", "D1d", "D6c", "D4d", "D1ab",	"D2", "D4c", "D1c", "C3f", "C3", "C50a", "C3fl", "C3ae", "C3fm", "C50q", "C3b", "C3h", "C3bj", "C3fo", "C3a", "C21", "C1", "C1c", "C1b", "C42.2", "C1br", "C1bh", "C1cb", "C72k", "C1C3", "C1c", "C1b", "C72k'", "C1w", "C1u", "C1bo", "C1y", "C15", "C15by", "A1", "A1ee", "C91",  "C15", "C15ek", "C15gy", "C15gk", "C15el", "B1", "B1a", "B1b", "B1g", "C42a", "C1", "C42.2", "C1j", "C1b", "C42as", "C42aa", "C42aq", "C1", "C42ar", "C42a", "C42b", "C1au", "C42g",	"C42h", "C42i",	"C1au", "A1", "A1bw", "A1bf", "A1bx", "C15by") 

DIV_list <- as.data.frame(DIV_list) %>%
  distinct()

differential_MED_list %>%
  filter(seq %in% DIV_list$DIV_list)

differential_MED_list %>%
  filter(str_detect(seq, "X"))

```


#Summary stats
## proportion of variants
```{r}


melt %>%
  filter(project == "c_v_e") %>%
  filter(Genus == "Cladocopium" | Genus == "Breviolum" | Genus == "Durusdinium" | Genus == "Symbiodinium") %>%
  filter(Abundance > 0) %>%
  subset(select = c(OTU, Abundance, Genus)) %>%
  group_by(Genus) %>% 
  summarize(Abundance_sum = sum(Abundance)) %>% 
  ungroup()


#C
7819177/9136596
#0.8558086

#D
1312411/9136596
#0.1436433

#A
3845/9136596
#0.0004208351

#B
1163/9136596
#0.0001272903


```
